tests/returntypes/test_torch_qnode_new.py

```diff
@@ -0,0 +1,2138 @@
+        )
+        expected_c = torch.tensor(
+            [0, -torch.sin(d[0] + w[0]) * torch.sin(d[1] + w[1]), torch.cos(d[1] + w[1])]
+        )
+
+        assert torch.allclose(grad[0], expected_w, atol=0.1)
+        assert torch.allclose(grad[1], expected_c, atol=0.1)
+
+        # test second-order derivatives
+        if diff_method == "parameter-shift" and max_diff == 2:
+            hessians = torch.autograd.functional.hessian(circuit, (d, w, c))
+
+            grad2_c = hessians[2][2]
+            assert torch.allclose(grad2_c, torch.zeros([3, 3], dtype=torch.float64), atol=0.1)
+
+            grad2_w_c = hessians[1][2]
+            expected = torch.tensor(
+                [
+                    [0, -torch.cos(d[0] + w[0]) * torch.sin(d[1] + w[1]), 0],
+                    [
+                        0,
+                        -torch.cos(d[1] + w[1]) * torch.sin(d[0] + w[0]),
+                        -torch.sin(d[1] + w[1]),
+                    ],
+                ]
+            )
+            assert torch.allclose(grad2_w_c, expected, atol=0.1)
+
+
+class TestSample:
+    """Tests for the sample integration"""
+
+    def test_sample_dimension(self):
+        """Test sampling works as expected"""
+        dev = qml.device("default.qubit", wires=2, shots=10)
+
+        @qnode(dev, diff_method="parameter-shift", interface="torch")
+        def circuit():
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.sample(qml.PauliZ(0)), qml.sample(qml.PauliX(1))
+
+        res = circuit()
+
+        assert isinstance(res, tuple)
+        assert len(res) == 2
+
+        assert tuple(res[0].shape) == (10,)
+        assert isinstance(res[0], torch.Tensor)
+
+        assert tuple(res[1].shape) == (10,)
+        assert isinstance(res[1], torch.Tensor)
+
+    def test_sampling_expval(self):
+        """Test sampling works as expected if combined with expectation values"""
+        shots = 10
+        dev = qml.device("default.qubit", wires=2, shots=shots)
+
+        @qnode(dev, diff_method="parameter-shift", interface="torch")
+        def circuit():
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.sample(qml.PauliZ(0)), qml.expval(qml.PauliX(1))
+
+        res = circuit()
+
+        assert len(res) == 2
+        assert isinstance(res, tuple)
+
+        assert isinstance(res[0], torch.Tensor)
+        assert res[0].shape == (shots,)
+        assert isinstance(res[1], torch.Tensor)
+        assert res[1].shape == ()
+
+    def test_counts_expval(self):
+        """Test counts works as expected if combined with expectation values"""
+        shots = 10
+        dev = qml.device("default.qubit", wires=2, shots=shots)
+
+        @qnode(dev, diff_method="parameter-shift", interface="torch")
+        def circuit():
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            return qml.counts(qml.PauliZ(0)), qml.expval(qml.PauliX(1))
```

Comment:
Out of scope: eventually it could be good to add more battle-testing to `qml.counts` to interfaces - there are variants where no op or wires are provided. Maybe something for a GitHub issue, just to have more coverage of edge cases.

Reply:
Yes wwe should do that after merging pytorch and jax-jit